Show deposits for fees and transactions

hasufell opened the pull request:
# Issue Number

ADP-470

Commit:
Show deposits for fees and transactions

hasufell commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -502,6 +502,8 @@ fromShelleyPParams bound pp = W.ProtocolParameters
         desiredNumberOfStakePoolsFromPParams pp
     , minimumUTxOvalue =
         minimumUTxOvalueFromPParams pp
+    , poolDeposit = poolDepositFromPParams pp
```
Add deposits to ProtocolParameters...

hasufell commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -746,6 +747,7 @@ data ApiTransaction (n :: NetworkDiscriminant) = ApiTransaction
     , withdrawals :: ![ApiWithdrawal n]
     , status :: !(ApiT TxStatus)
     , metadata :: !ApiTxMetadata
+    , deposits :: ![Quantity "lovelace" Natural]
```
adjust API types

KtorZ replied:
What do you mean ?

hasufell replied:
I meant this adjusted the API types :sweat_smile: ... these comments were meant as a walkthrough

hasufell commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -2070,9 +2075,13 @@ joinStakePool ctx currentEpoch knownPools pid poolStatus wid =
         return $ if isKeyReg
-            then Join pid
-            else RegisterKeyAndJoin pid
+            then (Join pid, dep)
```
I think this is a prior bug?

hasufell commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1421,7 +1421,12 @@ estimateFeeForDelegation ctx wid = db & \DBLayer{..} -> do
+    dep <- fmap keyDeposit $
+        withExceptT ErrSelectForDelegationNoSuchWallet
+            $ readWalletProtocolParameters @ctx @s @k ctx wid
+    estimateFeeForCoinSelection (if isKeyReg then getCoin dep else 0)
```
Interesting part: here we actually return some non-zero deposit

hasufell commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -2070,9 +2075,13 @@ joinStakePool ctx currentEpoch knownPools pid poolStatus wid =
-            then Join pid
-            else RegisterKeyAndJoin pid
+            then (Join pid, 0)
+            else (RegisterKeyAndJoin pid, dep)
```
Interesting part: here we actually return some non-zero deposit

KtorZ replied:
Why is that interesting :thinking: ?

hasufell replied:
:sweat_smile: because here happens something. Most of the PR is types fixing. I wanted to point out where actual business logic is changed/added.

hasufell commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -2070,9 +2075,13 @@ joinStakePool ctx currentEpoch knownPools pid poolStatus wid =
         return $ if isKeyReg
-            then Join pid
-            else RegisterKeyAndJoin pid
+            then (Join pid, 0)
```
I think Join and RegisterKeyAndJoin need to be flipped here? Is that a bug?

hasufell commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1197,7 +1197,8 @@ selectCoinsForJoin ctx knownPools getPoolStatus pid wid = do
     (utx, action, path) <- withWorkerCtx ctx wid liftE liftE $ \wrk -> do
-        action <- liftHandler
+        -- we never register stake key here, so deposit is irrelevant
+        (action, _) <- liftHandler
```
this is a little fishy... but currently there's no way to trigger a register stake key via this endpoint, so adding deposit to CoinSelection types seems excessive

KtorZ replied:
> there's no way to trigger a register stake key via this endpoint

I challenge this. When call for the very first time so that this is the first delegation for this wallet, it will result in a stake key registration. I do not understand your rationale here?

hasufell replied:
```hs
    coinSelections :: Server (CoinSelections n)
    coinSelections = (\wid ascd -> case ascd of
        (ApiSelectForPayment ascp) ->
            selectCoins shelley (delegationAddress @n) wid ascp
        (ApiSelectForDelegation (ApiSelectCoinsAction (ApiT action))) ->
            case action of
                Join pid -> selectCoinsForJoin @_ @()
                    shelley
                    (knownPools spl)
                    (getPoolLifeCycleStatus spl)
                    pid
                    (getApiT wid)
                -- if this would be enabled, we'd need to adjust selectCoinsForJoin
                -- to account for deposit
                RegisterKeyAndJoin _ -> throwError err400
                Quit -> selectCoinsForQuit @_ @() shelley wid
        )
```

`RegisterStakeKeyAndJoin` isn't allowed here.

KtorZ replied:
Yes, and this is why re-using this type was a bit hacky :upside_down_face: ... This does not translate what's happening in the actual underlying logic. Calls from `selectCoinsForJoin` can totally generate stake key registration with implied key deposits. 

I'd suggest defining a proper api type to reflect this and prevent someone else to be bitten by this. The API action that are supported are either Join/Quit and they map internally to Join / RegisterKeyAndJoin / Quit.

hasufell replied:
I've done:

1. Use a separate Api type for delegation action

```hs
data DelegationActionType = Join | Quit
    deriving (Eq, Generic, Show)
    deriving anyclass NFData

data ApiDelegationAction = ApiDelegationAction
    { action :: DelegationActionType
    , pool :: Maybe (ApiT PoolId)
    }
    deriving (Eq, Generic, Show)
    deriving anyclass NFData
```

2. add `deposits` field to coin selection

```hs
data ApiCoinSelection (n :: NetworkDiscriminant) = ApiCoinSelection
    { inputs :: !(NonEmpty (ApiCoinSelectionInput n))
    , outputs :: ![ApiCoinSelectionOutput n]
    , change :: ![ApiCoinSelectionChange n]
    , certificates :: Maybe (NonEmpty ApiCertificate)
    , deposits :: ![Quantity "lovelace" Natural]
    } deriving (Eq, Generic, Show)
      deriving anyclass NFData
```

hasufell commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -2049,7 +2054,7 @@ joinStakePool
     -> PoolLifeCycleStatus
     -> WalletId
-    -> ExceptT ErrJoinStakePool IO DelegationAction
+    -> ExceptT ErrJoinStakePool IO (DelegationAction, Coin)
```
Maybe a bit inelegant, but don't see a way to stuff it into DelegationAction and creating a separate type also seems excessive.

hasufell commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1457,11 +1459,12 @@ mkApiTransactionFromInfo
-    apiTx <- mkApiTransaction ti txid (drop2nd <$> ins) outs ws (meta, txtime) txmeta $
-        case meta ^. #status of
+    apiTx <- mkApiTransaction ti txid (drop2nd <$> ins) outs ws (meta, txtime) txmeta
+        (case meta ^. #status of
```
Funny thing... I couldn't move this stuff into a let or where binding without getting compile errors

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:
```diff
@@ -173,6 +173,8 @@ ProtocolParameters
     protocolParametersDecentralizationLevel Percentage      sql=decentralization_level
     protocolParametersDesiredNumberOfPools  Word16          sql=desired_pool_number
     protocolParametersMinimumUtxoValue      W.Coin          sql=minimum_utxo_value
+    protocolParametersPoolDeposit           W.Coin          sql=pool_deposit
```
This is superfluous for cardano-wallet, we do never register stake pools and won't do in any near nor distant future.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -502,6 +502,8 @@ fromShelleyPParams bound pp = W.ProtocolParameters
     , minimumUTxOvalue =
         minimumUTxOvalueFromPParams pp
+    , poolDeposit = poolDepositFromPParams pp
+    , keyDeposit = keyDepositFromPParams pp
```
stakeKeyDeposit would be a slightly better name, there are many things that can qualify as "key" in the application.

KtorZ commented on specifications/api/swagger.yaml:
```diff
@@ -1322,9 +1322,11 @@ components:
       properties:
         estimated_min: *transactionAmount
         estimated_max: *transactionAmount
+        deposit: *transactionAmount
```
For consistency, what about having a list of `deposits` here too ?

hasufell replied:
Is it possible to have theoretically multiple for a fee estimation?

KtorZ replied:
Theoretically yes. If we were to register multiple stake keys. Arguably, you could represent them as "one big deposit", but we're already taking the approach of showing them as multiple deposits in transactions.

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -2118,6 +2127,8 @@ data FeeEstimation = FeeEstimation
     , estMaxFee :: Word64
     -- ^ Most coin selections will result in a fee lower than this.
+    , estDeposit :: Word64
+    -- ^ Deposit if stake key was registered
```
This data-type feels a little bit _abused_. There's nothing of an estimation when it comes to stake key deposits. The deposit value is fixed by the network.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1461,11 +1463,12 @@ mkApiTransactionFromInfo
             InLedger -> #insertedAt
-            Expired  -> #pendingSince
+            Expired  -> #pendingSince)
+        []
```
To help readability, consider making this a named argument:

```
    let deposits = []
    apiTx <- mkApiTransaction ... deposits
```

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:
```diff
@@ -176,6 +176,7 @@ ProtocolParameters
     protocolParametersDecentralizationLevel Percentage      sql=decentralization_level
     protocolParametersDesiredNumberOfPools  Word16          sql=desired_pool_number
     protocolParametersMinimumUtxoValue      W.Coin          sql=minimum_utxo_value
+    protocolParametersKeyDeposit            W.Coin          sql=key_deposit
```
Sadly, this needs to be added as the very last. `persistent` does not "tolerate" adding new columns in the middle of existing declarations.

KtorZ reviewed (changes requested):
One thing important to notice is that, we are conflating the "key deposits" with the third fee component. The main reason for that was to accommodate both Jormungandr  and Cardano-node requirements: Jormungandr does not have a concept of "deposit", but instead, has an extra "certificate fee" which needs to be applied for each certificate. 

That is, we've defined transaction fees as: a + b*x + c*y where `x` and `y` depends on cardano-node / jormungandr. Now that jormungandr is gone, this is a little bit awkward because the key deposit is not exactly a fee. Plus, this PR now introduces yet-another-way to view the key deposit value (via the protocol parameters) which feels like the right way. We should consider revising the existing design to have one consistent view on key deposits. 

_I think_ the main change will be in the TransactionLayer initialization: 

```hs
    _initDelegationSelection
        :: FeePolicy
            -- Current fee policy
        -> DelegationAction
            -- What sort of action is going on
        -> CoinSelection
        -- ^ An initial selection where 'deposit' and/or 'reclaim' have been set
        -- accordingly.
    _initDelegationSelection (LinearFee _ _ (Quantity c)) = \case
        Quit{} -> mempty { reclaim = round c }
        Join{} -> mempty
        RegisterKeyAndJoin{} -> mempty { deposit = round c }
```

This `c` value would have to come from the protocol parameters more directly. Since the `c` component on the fee policy is now obsolete it makes sense to just remove it entirely and let the compiler guide us through places that are impacted.

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1879,7 +1885,7 @@ mkTxMeta interpretTime blockHeader wState tx cs expiry =
             = sum (fromIntegral . getCoin . coin . snd <$> (inputs cs))
             + sum (mapMaybe ourWithdrawal $ Map.toList $ withdrawals tx)
-            + fromIntegral (reclaim cs)
+            + fromIntegral (CoinSelection.deposit cs)
```
This is wrong. Deposits and reclaims aren't the same thing. Deposits counts towards the `output` balance actually.

hasufell replied:
Absolutely. I'm also not sure what was going on here.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1199,8 +1199,9 @@ selectCoinsForJoin ctx knownPools getPoolStatus pid wid = do
-    (utx, action, path) <- withWorkerCtx ctx wid liftE liftE $ \wrk -> do
-        action <- liftHandler
+    (utx, action, path, dep) <- withWorkerCtx ctx wid liftE liftE $ \wrk -> do
+        -- we never register stake key here, so deposit is irrelevant
```
This comment seems no longer accurate?

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1460,11 +1462,13 @@ mkApiTransactionFromInfo
             InLedger -> #insertedAt
-            Expired  -> #pendingSince
+            Expired  -> #pendingSince)
+        deposits
```
Small stylistic note: I'd suggest to position the deposit argument before the last argument for better readability. Arguably, that function already has way too many arguments and a record would be better in the first place.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -481,11 +481,23 @@ data ApiCertificate
     deriving (Eq, Generic, Show)
     deriving anyclass NFData
 
+data DelegationActionType = Join | Quit
```
*ApiDelegationActionType

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -481,11 +481,23 @@ data ApiCertificate
+
+data ApiDelegationAction = ApiDelegationAction
+    { action :: DelegationActionType
+    , pool :: Maybe (ApiT PoolId)
```
Why not: 

```hs
 data ApiDelegationActionType = Join PoolId | Quit 
```

to avoid situations where we have action = Quit and pool = Just {} , or reciprocally when we have action = Joint but pool = Nothing ... which wouldn't mean anything? Was this to work-around the JSON instance? Note that the same instance can be achieved with `Join PoolId | Quit` by using the following serialization options: 

```hs
apiDelegationActionTypeOptions :: Aeson.Options
apiDelegationActionTypeOptions = Aeson.defaultOptions
      { tagSingleConstructors = True
      , omitNothingFields = True
      , constructorTagModifier = camelTo2 '_'
      , fieldLabelModifier = camelTo2 '_' . dropWhile (== '_')
      , sumEncoding = TaggedObject
          { tagFieldName = "action"
          , contentsFieldName = "pool" 
          }
      }
```

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -497,6 +499,15 @@ migrateManually tr proxy defaultFieldValues =
+    addKeyDepositIfMissing conn = do
+        addColumn_ conn True (DBField ProtocolParametersKeyDeposit) value
+      where
+        value = T.pack $ show $ W.getCoin $ defaultKeyDeposit defaultFieldValues
```
```suggestion
        value = toText $ defaultKeyDeposit defaultFieldValues
```

KtorZ commented on lib/core/test/data/Cardano/Wallet/Api/ApiCoinSelectionTestnet0.json:
```diff
@@ -1,1961 +1,4893 @@
 {
-    "seed": 1426841331592113606,
+    "seed": -4641725427390187379,
```
That file is way too large, but that's tangent to this PR.

KtorZ commented on lib/core/test/data/Cardano/Wallet/Api/ApiDelegationAction.json:
```diff
@@ -0,0 +1,44 @@
+        {
+            "action": "quit",
+            "pool": "pool17qkm5ns0r2zl4gwrr4ne50kypxcvcdcq7956xu7jhm6f28pg5c8"
+        },
```
The meaning of this one (quit + pool) is dubious.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -1634,6 +1648,7 @@ instance Arbitrary (ApiTransaction t) where
             <*> genWithdrawals
             <*> pure txStatus
             <*> arbitrary
+            <*> arbitrary
```
Note: no need to generate infinitely large lists here.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Api/Server.hs:
```diff
@@ -236,16 +236,19 @@ server byron icarus shelley spl ntp =
+                (Join, Nothing) -> Handler $ throwE $ apiError err400 BadRequest $
+                    mconcat
+                    [ "Missing PoolId for Join action."
+                    ]
```
:upside_down_face:

hasufell replied:
?

KtorZ replied:
https://github.com/input-output-hk/cardano-wallet/pull/2327#discussion_r532434912

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -223,17 +222,8 @@ newTransactionLayer networkId = TransactionLayer
-        Join{} -> mempty
-        RegisterKeyAndJoin{} -> mempty { deposit = round c }
+        :: CoinSelection
+    _initDelegationSelection = mempty
```
I am not sure to understand this change? what happened to the reclaim and the deposit needed for the coin selection to run correctly?

hasufell replied:
I'm also wondering why I did that. But what's even more interesting is why the tests pass

https://github.com/input-output-hk/cardano-wallet/pull/2327/files#diff-0a90babe9bf43651092fb58d3a37bd24cc2ca0a6d63f948c6c64c738a27f5392L1392

hasufell replied:
Ah, now I remember... we removed the 3rd parameter for LinearFee https://github.com/input-output-hk/cardano-wallet/pull/2327/files#diff-45c6b9a78752ca17442ffe1ded898e5fd2f6931f95c9ae4ad21d30afcf0e2e1fL810 because that only existed because of jurmungandr as far as I remember the conversation

KtorZ replied:
Yes okay, the third parameter was removed from the linear fee, but it's because it now comes from the protocol parameters directly as the stake key deposit. The logic that applied here still applies however!

hasufell commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1875,11 +1881,13 @@ mkTxMeta interpretTime blockHeader wState tx cs expiry =
     let
         amtOuts =
             sum (mapMaybe ourCoins (outputs cs))
+            + fromIntegral (CoinSelection.deposit cs)
```
@KtorZ

KtorZ replied:
Yes? What/why is this change?

hasufell replied:
I wanted to know if this is correct, because of your comment https://github.com/input-output-hk/cardano-wallet/pull/2327#discussion_r532431872

KtorZ replied:
It is not. The current code is doing just fine :+1:

KtorZ replied:
(This calculation happens on the transaction outputs directly, so the deposit is already _implicitly_  included in the difference between inputs and outputs).

KtorZ replied:
Although, the point of this PR / Story is now to make it possible to identify whether the delta (in - out) is purely fee, or fee and deposits.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -1634,6 +1644,7 @@ instance Arbitrary (ApiTransaction t) where
             <*> genWithdrawals
             <*> pure txStatus
             <*> arbitrary
+            <*> resize 10 (listOf arbitrary)
```
> **resize:** _Overrides the size parameter. Returns a generator which uses the given size instead of the runtime-size parameter._

I am pretty sure that this is not what we want here :no_mouth:

hasufell replied:
Why?

This means that the *maximum* size of the list is going to be 10. You wanted the list to not be arbitrary length.

KtorZ replied:
It'll effectively work because of how `listOf` behaves. I just found it a bit misleading because `resize 10` means that the size parameter will always be `10` whereas QuickCheck typically grow the size parameter to be larger and larger and produce more and more complex cases. `resize 10` circumvent that by forcing the size to always be the same (which acts as a max-size for lists and most other structures, true, but misleading).

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1420,7 +1420,12 @@ estimateFeeForDelegation ctx wid = db & \DBLayer{..} -> do
+
+    dep <- fmap stakeKeyDeposit $
+        withExceptT ErrSelectForDelegationNoSuchWallet
+            $ readWalletProtocolParameters @ctx @s @k ctx wid
```
Note: this needs to be fed into the `initDelegationSelection` function to correctly initialize a coin selection with appropriate deposits (or reclaim). The only real difference with _before_ is that the deposit amount now comes directly from the "stakeKeyDeposit" protocol parameter, and not from the fee policy. Logic-wise, the rest should not change.

hasufell replied:
It is:

```hs
    let selectCoins = selectCoinsForDelegationFromUTxO @_ @t @k
            ctx utxo txp minUtxo dep action
```

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1420,7 +1420,12 @@ estimateFeeForDelegation ctx wid = db & \DBLayer{..} -> do
+    dep <- fmap stakeKeyDeposit $
+        withExceptT ErrSelectForDelegationNoSuchWallet
+            $ readWalletProtocolParameters @ctx @s @k ctx wid
+    estimateFeeForCoinSelection (if isKeyReg then Nothing else Just $ getCoin dep)
```
Note that passing this to the `estimateFeeForCoinSelection` function is a bit redundant with the coin selection itself. A coin selection is captured by:

```hs
data CoinSelection = CoinSelection
    { inputs  :: [(TxIn, TxOut)]
      -- ^ Picked inputs
    , withdrawal :: Word64
      -- ^ An available withdrawal amount, counting as an extra input
    , reclaim :: Word64
      -- ^ Claim back a deposit, counting as a an extra input
    , outputs :: [TxOut]
      -- ^ Picked outputs
    , change  :: [Coin]
      -- ^ Resulting changes
    , deposit :: Word64
      -- ^ A deposit counting as an extra output
    } deriving (Generic, Show, Eq)
```

So, the presence and full value of a deposit can be inferred from a `CoinSelection`. Therefore, a better approach which does not duplicate any logic with regards to deposits would be to change `feeBalance` to return not only a `Fee`, but also a deposit value and have that bubbling through `estimateFeeForCoinSelection`.

hasufell replied:
Doing it in `feeBalance` is quite intrusive. I believe it's best to do it in a lambda. So:

```hs
    estimateFeeForCoinSelection isKeyReg
        $ (\cs -> (Fee $ feeBalance cs, CoinSelection.deposit cs)) <$> selectCoins
```


----

Edit: I think it's more annoying than useful. `estimateFeeForCoinSelection` repeats the action, so stuffing it into a tuple with fee is just confusing here.

hasufell replied:
I think we should try again to move it out.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -574,6 +587,27 @@ migrateManually tr proxy defaultFieldValues =
+                    , ";"
+                    ]
+                Sqlite.step query *> Sqlite.finalize query
+            _ -> pure ()
```
Please stick to the existing patterns with regards to database migrations: 

a) Avoid hard-coded column and table names, and use instead the `DBField` data-type for existing database types.

b) Define proper logging messages depending on the situation (migration needed vs migration not needed)

c) Make sure to add a unit test for testing the migration (see examples in https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs#L250-L272) 

d) Make sure to start your migration query with an `isFieldPresent` request. The database may be older than you think.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -223,17 +223,17 @@ newTransactionLayer networkId = TransactionLayer
+        Quit{} -> mempty { reclaim = c }
         Join{} -> mempty
-        RegisterKeyAndJoin{} -> mempty { deposit = round c }
+        RegisterKeyAndJoin{} -> mempty { deposit = c }
```
:+1:

hasufell commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -589,24 +589,77 @@ migrateManually tr proxy defaultFieldValues =
+                case row of
+                    [PersistText t] -> do
+                        traceWith tr $ MsgManualMigrationNeeded field t
+                        [a, b, c] <- pure (T.splitOn " + " t)
```
Migrations don't need to be idempotent, do they? They're only ever run once, right?

Because if we would run this migration twice, this pattern match would fail.

hasufell commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -589,24 +589,77 @@ migrateManually tr proxy defaultFieldValues =
+                    _ -> fail "Couldn't query fee value!"
+      where
+        updateStakeKeyVal :: DBField -> Text -> IO ()
+        updateStakeKeyVal stakeField c = do
```
So I decided to also extract and update the stake key value here, which is the old 3rd parameter of `FeePolicy` (which is now removed).

Please confirm this is what we want.

hasufell commented on lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:
```diff
@@ -425,6 +431,43 @@ testMigrationPassphraseScheme = do
+                    readProtocolParameters wid
+        let migrationMsg = filter isMsgManualMigration logs
+        length migrationMsg `shouldBe` 1
+        (getFeePolicy . txParameters <$> pp') `shouldBe`
```
Here's the migration spec.

hasufell commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -589,24 +589,77 @@ migrateManually tr proxy defaultFieldValues =
+                    -- this invariant must hold due to the parent context
+                    fail "ProtocolParameters table missing"
+                ColumnMissing -> do
+                    traceWith tr $ MsgManualMigrationNotNeeded stakeField
```
Questionable, but well...

KtorZ replied:
Actually, a migration is needed. And that migration is `addKeyDepositIfMissing`. Which is a bit unfortunate, because now, if we compose the two actions above (`addKeyDepositIfMissing` and `updateFeeValue`), we end up in a situation where the column is created with a possibly wrong value...

I'd suggest adding the `key_deposit` column here, instead of `addKeyDepositIfMissing`.

hasufell commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -589,24 +589,77 @@ migrateManually tr proxy defaultFieldValues =
-                (a:b:_) <- pure (T.splitOn " + " t)
-                let newVal = a <> " + " <> b
+        let field = (DBField ProtocolParametersFeePolicy)
+            defaultValue = "0 + 0x"
```
I don't see any other reasonable default value here.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:
```diff
@@ -272,6 +274,10 @@ spec = do
 
+        it "'migrate' db with new FeePolicy" $
+            testMigrationUpdateFeeValue @ShelleyKey
+                "shelleyAccountingStyle-v2020-10-13.sqlite"
```
Lazy or smart. I can't tell :grimacing:

hasufell replied:
Haskellers know it can be both :grin:

hasufell commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -639,17 +638,26 @@ migrateManually tr proxy defaultFieldValues =
                     fail "ProtocolParameters table missing"
                 ColumnMissing -> do
-                    traceWith tr $ MsgManualMigrationNotNeeded stakeField
+                    addColumn_ conn True stakeField c
```
Here's the sweet juice :candy:

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1447,8 +1449,10 @@ mkApiTransactionFromInfo
+    let deposits = []
+    apiTx <- liftIO $ mkApiTransaction ti txid (drop2nd <$> ins) outs ws (meta, txtime) txmeta
+        deposits
+        $ case meta ^. #status of
```
This is no more readable than before. It's actually worse. The 80-column max-length rule should apply, but not be detrimental to readability.  If it goes slightly beyond 80 chars, we can live with it.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -569,6 +571,102 @@ migrateManually tr proxy defaultFieldValues =
+            value = toText $ defaultKeyDeposit defaultFieldValues
+
+        updateStakeKeyVal :: DBField -> Text -> IO ()
+        updateStakeKeyVal stakeField c = do
```
What about using a single function to cover all cases? Especially because they have quite different names although are acting on exactly the same table and column. `updateStakeKeyVal` should be enough in other cases too.

hasufell replied:
I tried changing

```hs
    addColumn
        :: Sqlite.Connection
        -> Bool
        -> DBField
        -> Text
        -> Maybe (IO ()) -- ^ action to run when column is present
        -> IO SqlColumnStatus
    addColumn conn notNull field value updateCB = do
        isFieldPresent conn field >>= \st -> st <$ case st of
            TableMissing ->
                traceWith tr $ MsgManualMigrationNotNeeded field
            ColumnMissing -> do
                traceWith tr $ MsgManualMigrationNeeded field value
                query <- Sqlite.prepare conn $ T.unwords
                    [ "ALTER TABLE", tableName field
                    , "ADD COLUMN", fieldName field
                    , fieldType field, if notNull then "NOT NULL" else ""
                    , "DEFAULT", value
                    , ";"
                    ]
                _ <- Sqlite.step query
                Sqlite.finalize query
            ColumnPresent ->
                maybe (traceWith tr $ MsgManualMigrationNotNeeded field) id updateCB
```

but I think it makes it more convoluted. Although some code is shared here, it doesn't improve understanding imo.

hasufell replied:
I improved the names a bit.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -1328,26 +1427,29 @@ mkProtocolParametersEntity
 mkProtocolParametersEntity wid pp =
-    ProtocolParameters wid fp (getQuantity mx) dl desiredPoolNum minUTxO epochNo
+    ProtocolParameters wid fp (getQuantity mx) dl desiredPoolNum minUTxO
+        keyDep epochNo
```
This is no more readable than before.

KtorZ reviewed (changes requested):
we must keep track of deposit as part of the transaction metadata so that the deposit can be reported correctly even once we are no longer aware of the initial coin selection.

Commit:
Store deposit in database and return in TransactionInfo

Commit:
Add integration test for checking deposits

This ought to fail right now.

hasufell commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1899,6 +1899,7 @@ mkTxMeta ti' blockHeader wState tx cs expiry =
                 , blockHeight = blockHeader ^. #blockHeight
                 , amount = Quantity $ distance amtInps amtOuts
                 , expiry = Just expiry
+                , deposit = Just . Quantity . Coin $ (cs ^. #deposit)
```
This should do the trick... however... it doesn't. It still doesn't end up in the database. I couldn't figure out until now why. I followed the control flow and it should be hitting the database.

KtorZ replied:
This should be `Nothing` if `cs ^. #deposit == 0`

hasufell replied:
https://github.com/input-output-hk/cardano-wallet/blob/d4187a713364dca2f00a073338cdc65d37693d36/lib/core/src/Cardano/Wallet/Api/Server.hs#L1495-L1553
This is where the `txMeta` gets updated and the handler indeed returns a deposit. At the same time the handler also transmits this txMeta to the database in https://github.com/input-output-hk/cardano-wallet/blob/d4187a713364dca2f00a073338cdc65d37693d36/lib/core/src/Cardano/Wallet.hs#L1924-L1941

But it never shows up in the database.

KtorZ replied:
What do you mean "it never shows up" ?

hasufell replied:
> What do you mean "it never shows up" ?

The database field is `null`.

And I found the culprit: https://github.com/input-output-hk/cardano-wallet/blob/d4187a713364dca2f00a073338cdc65d37693d36/lib/core/src/Cardano/Wallet/Primitive/Model.hs#L383-L392

I don't see how to easily access `TransactionInfo` in that function without database access.

KtorZ replied:
The model needs to be able to reconstruct the deposit's amount from the certificates present in the transaction and the active protocol parameters. 

The main issue here is to know what is the right value of the deposit _at the time the transaction was inserted_. We should be able to land back on our feet by checking the delta between inputs and outputs and the explicit transaction fees. With these, it becomes a simple equation with a single unknown.

hasufell commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:
```diff
@@ -215,11 +215,24 @@ spec = describe "SHELLEY_STAKE_POOLS" $ do
             , expectField (#direction . #getApiT) (`shouldBe` Outgoing)
+            , expectField #deposits (`shouldBe` [Quantity 1000000])
             ]
+        eventually "Wallet has joined pool and deposit info persists" $ do
```
Here's the new test

KtorZ commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:
```diff
@@ -215,11 +215,24 @@ spec = describe "SHELLEY_STAKE_POOLS" $ do
+                [ expectResponseCode HTTP.status200
+                , expectField (#status . #getApiT) (`shouldBe` InLedger)
+                , expectField (#direction . #getApiT) (`shouldBe` Outgoing)
+                , expectField #deposits (`shouldBe` [Quantity 1000000])
```
Please avoid the hard-coded value and use `depositAmt`

KtorZ commented on lib/core-integration/src/Test/Integration/Scenario/CLI/Shelley/Transactions.hs:
```diff
@@ -125,7 +125,7 @@ spec = describe "SHELLEY_CLI_TRANSACTIONS" $ do
         args <- postTxArgs ctx wSrc wDest amt Nothing Nothing
         Stdout feeOut <- postTransactionFeeViaCLI ctx args
-        ApiFee (Quantity feeMin) (Quantity feeMax) <- expectValidJSON Proxy feeOut
+        ApiFee (Quantity feeMin) (Quantity feeMax) [Quantity 0] <- expectValidJSON Proxy feeOut
```
That seems wrong. If there's no deposit, then there's no deposit.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:
```diff
@@ -104,6 +104,7 @@ TxMeta
     txMetaAmount            Natural             sql=amount
     txMetaData              W.TxMetadata Maybe  sql=data
     txMetaSlotExpires       SlotNo Maybe        sql=slot_expires
+    txMetaDeposit           W.Coin Maybe        sql=deposit
```
For consistency and to avoid ambiguity, can we make it `sql=key_deposit`

KtorZ committed:
review / simplify database migration for stake key deposit and fee policy.

KtorZ committed:
store and return explicit tx fee instead of deposits

  This is a little turn-around compared to the original approach but,
  since we need to be able to also recover fee or deposit in discovered
  transaction it is much easier to store and look at fee than deposits
  (which are not explicitly mentioned in transactions).

  Deposits can then be extrapolated from the delta between inputs and
  outputs and actual fees.

  Anyway, storing fees or deposits requires an important and non-trivial
  database migration.

hasufell commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -671,106 +678,81 @@ migrateManually tr proxy defaultFieldValues =
-                    xs -> fail ("Unexpected row result when querying fee value: " <> show xs)
+                case filter (/= PersistNull) row of
+                    [PersistText t] -> case T.splitOn " + " t of
+                        [a,b,c] -> do
```
This is different than before and less permissive.

KtorZ committed:
add migration for 'fee' in the tx_meta table

  This migration is really non-trivial and I am a little worried about
  what would happen if it was interrupted in the middle of it. We should
  make sure to clearly mention in the release notes that the next
  version of cardano-wallet may take some time on its first start due to
  heavy database migration.

  Codewise, it's been tested on rather old database (which hadn't even
  knowledge of the tx_withdrawal table) and it seems to work pretty
  well.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:
```diff
@@ -296,6 +305,65 @@ spec = parallel $ do
+                (LinearFee (Quantity 155381.0)  (Quantity 44.0))
+                (Coin 2000000)
+
+        it "'migrate' db to add fees to transactions" $
```
@piotr-iohk do you see any other interesting transaction from this database maybe that would be worth controlling?

piotr-iohk replied:
Maybe two more with a bit more inputs / outputs:

`99907bf6ac73f6fe6fe25bd6b68bae6776425b9d15a7c46c7a49b85b8b03f291`:

10 inputs
10 outputs
withdrawal = 565251

`15940a7c1df8696279282046ebdb1ee890d4e9ac3c5d7213f360921648b36666`:

1 input
33 outputs
withdrawal = 561120

KtorZ replied:
> Maybe two more with a bit more inputs / outputs:

Good idea!

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:
```diff
@@ -313,6 +381,62 @@ sqliteSpecRnd = do
+                Nothing ->
+                    fail $ "tx not found: " <> T.unpack (toText txid)
+                Just TransactionInfo{txInfoMeta} ->
+                    (txInfoMeta ^. #fee) `shouldBe` Just (Quantity expectedFee)
```
@piotr-iohk anything else you could think of that the migrated table should satisfy. In the end, the change is only operated on the `fee` column, so there isn't much more we can test for this particular column I believe...

piotr-iohk replied:
Perhaps validate also the calculated deposit? It will be 0 for all cases but one in current set of cases.

KtorZ committed:
make sure to attach resolved inputs to transactions when constructing them

  A year-old bug, where we would not return resolved inputs of
  transactions constructed in the API. When listed, transactions would
  be reported correctly as the history would be reconstructed from joint
  tables. But when first constructed, we need to make sure to re-use the
  fully qualified coin-selection inputs.

KtorZ committed:
fix overly large generated ApiCoinSelection data-type in JSON encoding tests.

KtorZ reviewed (approved):
Ok. @jonathanknowles it'd be good if you can do a quick sanity check before merging. Testing is pretty good on these areas and already caught many things (even a 1-year-old bug!) but another quick look won't hurt.

hasufell commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2035,6 +2049,61 @@ mkApiTransaction ti txid ins outs ws (meta, timestamp) txMeta setTimeReference =
+        -- For outgoing transactions however, the total in is garanteed to be
+        -- greater or equal to totalOut; any reminder is actually a
+        -- deposit. Said differently, if totalIn > 0, then necessarily 'fee' on
+        -- metadata should be 'Just{}'
```
```suggestion
        -- For outgoing transactions however, the totalIn is guaranteed to be
        -- greater or equal to totalOut; any remainder is actually a
        -- deposit. Said differently, if totalIn > 0, then necessarily 'fee' on
        -- metadata should be 'Just{}'.
```

hasufell commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -696,9 +709,171 @@ migrateManually tr proxy defaultFieldValues =
+    -- - Fees
+    -- - Total deposits if any
+    --
+    -- To substract deposit values from fees, we consider that any transaction
```
Is this a reasonable assumption? I don't have an intuition for this, but I also don't see a hard justification for it.

KtorZ replied:
It is, on mainnet, testnet and staging networks at least. Surely, if one were to set very very small fee, it wouldn't be but such network doesn't exist yet and if it were to be created it wouldn't require any database migration in the first place. 

The rationale for this justification is that we know that key registration transaction coming from the wallet have by construction one single output and only need to cover fees. So they typically have a small number of inputs as well. A single output transaction is about 0.17 Ada, 0.20 Ada if you're unlucky and need several inputs. An input is ~40 bytes, so 1800 Lovelace on public networks. So you would need more than 1100 inputs for this assumption to become false (with 1100 inputs one could create a transaction that is more expensive than the key deposit, while still not be a key registration transaction). cardano-wallet would certainly not have constructed such transaction.

hasufell commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -696,9 +709,171 @@ migrateManually tr proxy defaultFieldValues =
+    --
+    -- Note that ideally, we would do this in a single `UPDATE ... FROM` query
+    -- but the `FROM` syntax is only supported in SQLite >= 3.33 which is only
+    -- supported in the latest version of persistent-sqlite (2.11.0.0). So
```
There are many reasons to abandon our fork of persistent-sqlite. This is another.

KtorZ replied:
Or to get away from persistent altogether :)

hasufell replied:
True. I think ORMs don't have much value, unless you assume inexperienced developers. Serialisation can be easily tested and optimizing for performance becomes less of a hassle. But that would probably be a massive patch.

KtorZ replied:
> be a massive patch.

Not _much_ in the end. We used an indirection from the start, so it is a matter of re-implementing the [DBLayer](https://github.com/input-output-hk/cardano-wallet/blob/c955ca65309e1f74578e517792235816447ad19b/lib/core/src/Cardano/Wallet/DB.hs#L119) in terms of _something else_. We already have two implementations of it (SQLite with persistent & MVar)

hasufell commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -696,9 +709,171 @@ migrateManually tr proxy defaultFieldValues =
         value = T.pack $ show $ Seq.getAddressPoolGap Seq.defaultAddressPoolGap
 
+    -- This migration is rather delicate. Indeed, we need to introduce an
+    -- explicit 'fee' on known transactions, so only do we need to add the new
```
```suggestion
    -- explicit 'fee' on known transactions, so not only do we need to add the new
```

hasufell commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -696,9 +709,171 @@ migrateManually tr proxy defaultFieldValues =
+    -- transactions without having access to the entire ledger (we do not know
+    -- the _amount_ from inputs of incoming transactions). Therefore, by
+    -- convention it has been decided that incoming transactions will have fee
+    -- equals to 0.
```
Do we know if that could cause issues further down the line? Is it possible to use another magic value (such as `-1`) or make it nullable?

KtorZ replied:
No no. `0` is the value we want here for it conveys the right intuition: if I didn't make that transaction, from my perspective fees are 0. Since wallets are already user-centric (which is the reason why we can't know the amount any arbitrary inputs, but only those belonging to the wallet), it's a reasonable justification.

hasufell commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -696,9 +709,171 @@ migrateManually tr proxy defaultFieldValues =
+    -- Note that ideally, we would do this in a single `UPDATE ... FROM` query
+    -- but the `FROM` syntax is only supported in SQLite >= 3.33 which is only
+    -- supported in the latest version of persistent-sqlite (2.11.0.0). So
+    -- instead, we queries all transactions which require an update in memory,
```
```suggestion
    -- instead, we query all transactions which require an update in memory,
```

hasufell reviewed (commented):
Generally, I'd prefer if we went the deposit route, but I cannot really judge how much work that would be. We definitely need to open a ticket for this though, since it's basically technical debt.

KtorZ committed:
rework added code after multi-asset merge request (e.g. getCoin -> unCoin)

KtorZ committed:
fix worker registry causing SQLite wreckage on 'before' failure.

  If anything went wrong during the 'before' step when acquiring resources for a worker, we would report the resource as acquired and continue as normal. The error would be swallowed and completely silent and the server would simply grant us with an infamous 'StatementAlreadyFinalized' error on the next database call. This is a bit lame since we had already all the logic in place to correctly handle gracefully resource acquisation failures in the worker registry. Now fixed, the error is properly logged and the resource NOT acquired; the resulting worker is put in a 'dead' state. Note that failures of this kind are highly unexpected and should happen under no circumstances.

KtorZ committed:
make 'chunkSize' dependent of the underlying database schema

  After adding a new field on the 'TxMeta' table, I ended up hitting the SQLITE_MAX_VARIABLE_NUMBER=999 again on some scenarios. Indeed, we had hardcoded a chunksize of '100', and with a TxMeta with now 10 columns, any big query would become too large. Instead of reducing the chunksize to yet-another-arbitrary value and let someone else be bitten by this in the future (although the error reporting should now better thanks to the previous commit), I've reworked a bit the 'dbChunked' helper to infer the right chunksize from the model table itself such that adding more columns would automatically be accounted for.

rvl committed:
integration tests: Remove duplicate function, remove an error call

It's possible to throw an exception -- this function can run in IO.

rvl committed:
Adjust worker registry for unliftio/safe-exceptions

Async exceptions are not caught by `catch`.

But anyway, `catch` should be used for error recovery, not cleanup.

The function `withException` can handle cleanup if the "workerBefore"
action fails/is cancelled.

It's also possible to let exceptions in the worker thread bubble up
until they hit `forkFinally`.

I renamed a few variables to help ease understanding.

rvl committed:
Handle SIGINT at the top level, instead of in every worker thread

This removes differentiation between worker "thread killed" vs
"program interrupted". Treat them both the same as
"cancelled". Cleanup actions should be done in either case.

At the top level of the program, log the UserInterrupt exception which
results from Ctrl-C being pressed by the user.

KtorZ commented on lib/core/src/Cardano/Wallet/Registry.hs:
```diff
@@ -195,10 +193,9 @@ defaultWorkerAfter
-        _ -> MsgUnhandledException $ pretty $ show e
+    Left e -> if isSyncException e
+        then MsgThreadCancelled
+        else MsgUnhandledException $ pretty $ show e
```
:+1: ok

KtorZ replied:
Ah, no. Not okay. The `then` and `else` clauses are mixed up.

KtorZ replied:
Fixed in: https://github.com/input-output-hk/cardano-wallet/compare/fe58426a3cd9609800c4eac9b5662f74bdb2a4dd..db3dc33736b7dd6424bb2afa32a88def19bacd20
